lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -3249,6 +3249,11 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
 
   auto checkElementCast = [&](Type fromElt, Type toElt,
                               CheckedCastKind castKind) -> CheckedCastKind {
+    // Let's not emit diagnostic when the element type is erased because
+    // we can't statically know if element is convertible.
+    if (fromElt->isAny() || toElt->isAny())
+      return castKind;
```

Comment:
Note that this doesn't handle other cases such as `[AnyObject]` to `[NSString?]`. Am I right in saying that we only need this logic here to work around the fact that `typeCheckCheckedCast` will return an unresolved result for any cast to a more optional type? If so, I wonder if we should introduce a new `CheckedCastContextKind` for the element type of a collection and allow such to-more-optional casts for it? That should allow us to handle the more general cases such as with `AnyObject`.

Reply:
Yes, that is true ... I think a new `CheckedCastContextKind` will make this simpler so we can bail early with a value cast to or from Any and AnyObject in type check checked cast? Or we can just check `AnyObject` here too which is the centralized place that handles element for array, set, dictionary. WDYT?

Reply:
@LucianoPAlmeida Sorry, my suggestion could have been clearer :) What I'm suggesting is adding a bit of logic to the case where we fail to unwrap `fromType` here:

https://github.com/apple/swift/blob/440cd0bcfcde40bad45edc07feaa765532f5cea6/lib/Sema/TypeCheckConstraints.cpp#L3093-L3097

And proceeding with the unwrap of `toType` without returning `CheckedCastKind::Unresolved` or emitting an error if the `CheckedCastContextKind` is the new collection element context kind.

By continuing with the main logic of `typeCheckCheckedCast`, I believe the `Any` and `AnyObject` cases should be already handled properly, so we shouldn't need to add any additional checks for them :)

Reply:
Ahh, I see ... thanks @hamishknight, to me it just looking like a general case of Any and AnyObject. Ok I'm on it :)

Reply:
Yeah, my main concern with adding special cases for `Any` and `AnyObject` in the element coercion logic is that there may be other special cases that we haven't thought of that the main logic of `typeCheckCheckedCast` already handles

Reply:
For example :)

```
protocol P {}
struct S : P {}

func foo(_ x: [P]) {
  let _ = x as? [S?]
}
```

Reply:
uhum, I hadn't thought about `AnyObject` before, but to me, the case with any it was simpler because we can't assume anything statically so seemed fine to just special case it

Reply:
Ack, sorry, I just remembered SR-6192. It's actually possible for any collection-to-collection cast to succeed at runtime as an empty array/set/dictionary can be converted to any element type. Unfortunately this means that we can't produce any unrelated warnings, even for a cast from e.g `[Int]` to `[String]`.

The good news at least is the fix is even simpler – we just need to return `castKind` for `case CheckedCastKind::Unresolved:` in `checkElementCast`. No additional `CheckedCastContextKind` required :)

Could you also add both the above test case of `[P]` to `[S?]` and one testing `[Int]` to `[String]` and make sure we don't produce warnings?

Reply:
@hamishknight I think I'm going to skip those two (SR-7187, SR-6112) issues involving array elements in this PR, because I think we have to put a bit o more thought on that specific case, where some cases like involving literals, e.g. [1] as? [String] we should still warn... so as this PR has already grown a bit, I think is going to make it easy to reason about and even for review if we handle this separately :)

Reply:
@LucianoPAlmeida Okay no worries, sounds good!